task/validate-api/build.rs

```diff
@@ -24,16 +24,59 @@ fn write_pub_device_descriptions(
     let file = std::fs::File::create(dest_path)?;
     let mut file = std::io::BufWriter::new(file);
 
+    writeln!(
+        file,
+        "pub const MAX_ID_LENGTH: usize = {};",
+        SpComponent::MAX_ID_LENGTH,
+    )?;
+
     writeln!(
         file,
         "pub const DEVICES_CONST: [DeviceDescription; {}] = [",
         devices.len()
     )?;
 
-    for dev in devices {
+    //
+    // If a device in the TOML has no refdes, has the same refdes and suffix as
+    // another device, or produces a refdes-and-suffix string that is longer
+    // than the max component ID length, we will generate code that will not
+    // compile, so these errors are all fatal. However, as we loop over devices,
+    // we'll just log them and keep going, so that we can tell the user about
+    // *all* the bad devices in the config file, rather than bailing out at the
+    // first one. At the end, we return an error if there were any bad devices.
+    // This way, you don't have to fix one issue and recompile in order to
+    // discover the next error.
+    //
+    let mut missing_ids = 0;
+    let mut duplicate_ids = 0;
+    let mut ids_too_long = 0;
+    let mut id2idx = std::collections::BTreeMap::new();
```

Comment:
Add a comment here that we're using a `BTreeMap` to ensure that stuff is sorted?